src/sinks/util/auto_concurrency/controller.rs

```diff
@@ -182,14 +200,16 @@ impl<L> Controller<L> {
                     concurrency: inner.current_limit as u64,
                     reached_limit: inner.reached_limit,
                     had_back_pressure: inner.had_back_pressure,
-                    current_rtt: Duration::from_secs_f64(current_rtt),
+                    current_rtt: current_rtt.map(Duration::from_secs_f64),
                     past_rtt: Duration::from_secs_f64(past_rtt),
                 });
             }
 
             // Reset values for next interval
-            let new_avg = inner.past_rtt.update(current_rtt);
-            inner.next_update = now + Duration::from_secs_f64(new_avg);
+            if let Some(current_rtt) = current_rtt {
+                inner.past_rtt.update(current_rtt);
+            }
+            inner.next_update = now + Duration::from_secs_f64(inner.past_rtt.average());
```

Comment:
This seems like it can end up being `inner.next_update = now + Duration::from_secs_f64(0.0);` - in case the `EWMA` didn't update.

It's a bit hard to follow along that are the consequences of `current_rtt.average()` being `None`.

Reply:
No, if the `past_rtt` wasn't updated, it still has a value since this part of the condition cannot be entered unless `past_rtt` had a value. I agree this is unclear so I'll rework the logic to make it clearer.